### PR TITLE
Remove eval from base Editor

### PR DESCRIPTION
**Pull request by corranwebster**
The first in a series of PRs to replace #684 

This removes the use of eval in the `Editor` class and generally makes the synchronization code much clearer.

The code being changed has fairly comprehensive tests in `traitsui/tests/test_editor.py`.

### CODE REVIEW

File: traitsui/editor.py [hunk 1]
```diff
@@ -129,6 +135,15 @@ class Editor(HasPrivateTraits):
     #: The current editor invalid state status:
     invalid = Bool(False)
 
+    #: A set to track values being updated to prevent infinite recursion.
```
Comment (mdickinson) [inline traitsui/editor.py]: These are attributes that are already being used, but not set. Is that correct?

File: traitsui/editor.py [hunk 2]
```diff
@@ -129,6 +135,15 @@ class Editor(HasPrivateTraits):
...
+    _no_trait_update = Set()
+
+    #: A list of all values synchronized to.
```
Comment (mdickinson) [inline traitsui/editor.py]: Please could you indicate either in the comment or in the trait type definition what the contents of the tuples are? (What their expected length is, and what the contents of the fields are.)

File: traitsui/editor.py [hunk 3]
```diff
@@ -129,6 +135,15 @@ class Editor(HasPrivateTraits):
     #: The current editor invalid state status:
     invalid = Bool(False)
 
+    #: A set to track values being updated to prevent infinite recursion.
+    _no_trait_update = Set()
```
Comment (mdickinson) [inline traitsui/editor.py]: From the code, it looks as though this is a set of strings. Is that correct? If so, please could you add that information to the traits doc-comment?

File: traitsui/editor.py [hunk 4]
```diff
@@ -377,32 +392,86 @@ def get_undo_item(self, object, name, old_value, new_value):
...
+        Returns
+        -------
+        object, name, getter : any, str, callable
+            The object from the context, the (extended) name of the
+            attributes holding the value, and a callable which gets the
+            current value from the context.
         """
-        col = name.find(".")
-        if col < 0:
+        parts = name.split(".", 1)
```
Comment (mdickinson) [inline traitsui/editor.py]: Could also use `object_name, _, name = name.partition(".")` (and then test whether `name` is empty)

File: traitsui/editor.py [hunk 5]
```diff
@@ -377,32 +392,86 @@ def get_undo_item(self, object, name, old_value, new_value):
...
+                mode = trait.sync_value
             if isinstance(value, ContextValue):
-                self_trait = self.trait(name)
                 self.sync_value(
                     value.name,
                     name,
-                    self_trait.sync_value or trait.sync_value,
-                    self_trait.is_list is True,
+                    mode,
+                    bool(self_trait.is_list),
```
Comment (mdickinson) [inline traitsui/editor.py]: Is the `bool` call necessary? Does the `sync_value` method expect a strict boolean?

File: traitsui/editor.py [hunk 6]
```diff
@@ -445,146 +512,121 @@ def sync_value(
...
+    def _bind_from(self, key, user_object, xuser_name, editor_name, is_list):
+        """ Bind trait change handlers from a user object to the editor.
+
+        Parameters
+        ----------
+        key : str
+            The key to use to guard against recursive updates.
+        user_object : object
+            The extended name of the trait to be used on the user object.
+        editor_name : string
```
Comment (mdickinson) [inline traitsui/editor.py]: Missing a description of the `xuser_name` parameter, and the description for `user_object` doesn't look right

File: traitsui/editor.py [hunk 7]
```diff
@@ -445,146 +512,121 @@ def sync_value(
...
+    def _bind_to(self, key, user_object, xuser_name, editor_name, is_list):
+        """ Bind trait change handlers from a user object to the editor.
+
+        Parameters
+        ----------
+        key : str
+            The key to use to guard against recursive updates.
+        user_object : object
+            The extended name of the trait to be used on the user object.
+        editor_name : string
```
Comment (mdickinson) [inline traitsui/editor.py]: Also missing `xuser_name`, and `user_object` doesn't look right

File: traitsui/editor.py [hunk 8]
```diff
@@ -445,146 +512,121 @@ def sync_value(
...
+        """
+
+        def editor_trait_modified(new):
+            if key not in self._no_trait_update:
+                with self.no_trait_update(key), self.raise_to_debug():
+                    xsetattr(user_object, xuser_name, new)
+
+        self.on_trait_change(editor_trait_modified, editor_name)
+
+        if self._user_from is None:
```
Comment (mdickinson) [inline traitsui/editor.py]: I'll refrain from commenting on this, since it's unchanged by the PR ...

D'oh!

File: traitsui/editor.py [hunk 9]
```diff
@@ -598,3 +640,30 @@ def save_prefs(self):
         """ Returns any user preference information associated with the editor.
         """
         return None
+
+    # Utility context managers ----------------------------------------------
+
+    @contextmanager
+    def no_trait_update(self, name):
+        """ Context manager that blocks updates from the named trait. """
+        if self._no_trait_update is None:
```
Comment (mdickinson) [inline traitsui/editor.py]: This branch looks unnecessary: I don't think `_no_trait_update` can ever be `None`.

File: traitsui/editor.py (hunk 1 above)
Comment (corranwebster) [reply traitsui/editor.py]: This is the set of attributes on the object which have an active trait change in progress that is updating the editor.

File: traitsui/editor.py (hunk 8 above)
Comment (mdickinson) [reply traitsui/editor.py]: But it does look as though all the `is None` and `is not None` checks on `self._user_to` and `self._user_from` should be unnecessary, now that we have the trait defaults.

File: traitsui/editor.py (hunk 5 above)
Comment (corranwebster) [reply traitsui/editor.py]: Not necessary, but I think the value returned by `is_list` will frequently be `None` due to the way that trait metadata works, and given that the line was previously `self_trait.is_list is True` this is a more modern and safer rendering of the original code.

It's likely to save debugging time if someone runs into a problem with the code.

File: traitsui/editor.py (hunk 4 above)
Comment (corranwebster) [reply traitsui/editor.py]: I think `rpartition` is better given the subsequent code.

File: traitsui/editor.py (hunk 8 above)
Comment (corranwebster) [reply traitsui/editor.py]: This was `HasPrivateTraits` nonsense, but now that the traits are properly defined, it's not needed.

File: traitsui/editor.py [hunk 10]
```diff
@@ -552,8 +551,10 @@ def _bind_from(self, key, user_object, xuser_name, editor_name, is_list):
         key : str
             The key to use to guard against recursive updates.
         user_object : object
+            The object in the TraitsUI context that is being bound.
+        xuser_name: : str
             The extended name of the trait to be used on the user object.
-        editor_name : string
+        editor_name : str
```
Comment (mdickinson) [inline traitsui/editor.py]: There are other uses of `string` as type descriptions in this module, but changing all those back to `str` in this PR would be feature creep, I think.

Comment (corranwebster) [reply traitsui/editor.py]: Fixed in #686

**Pull request merged**